Homework02/perceptron.py

```diff
@@ -0,0 +1,51 @@
+"""
+perceptron class
+
+Created: 01.11.21, 20:04
+
+Author: LDankert
+"""
+import numpy
+import numpy as np
+
+
+class Perceptron:
+
+    # Constructor
+    def __init__(self, input_units):
+        self.weights = np.random.randn(input_units)
+        self.bias = np.random.randn()
+        self.alpha = 1
+        self.activation = 0
+
+    # Calculates the activation of the perceptron, takes 1d array
+    def forward_step(self, inputs):
+        # raise an exception if the number of inputs not match number of weights
+        if np.size(inputs) != np.size(self.weights):
+            raise Exception ("Inputs number doesn't match weight numbers")
+        else:
+            # multiply the weights with corresponding inputs
+            activation = np.sum((inputs*self.weights))
+            # adding the bias to the activation
+            activation += self.bias
+            self.activation = sigmoid(activation)
+
+    # Updates the parameters of the perceptron
+    def update(self, delta):
+        gradient = delta * self.activation
+        self.weights = self.weights - self.alpha * gradient
```

Comment:
If we apply the previous steps that would mean the bias is already a part of the weight matrix. Only these two lines would be needed:
`gradients = delta * self.activation`
`self.weights = self.weights - self.alpha * gradients`

Reply:
I think its not really. Ofc it is in the net input and therefore in the activation, but just the value, not the bias itself. So we have to update that as well to make sure the next time we calc the activation the bias has again a updated value

Reply:
Isn't it that we add bias weights to the weight matrix itself so we only need to update the latter one. The bias weights are, sort of, multiplied by the default input of 1. It's mentioned in the "Matrix Notation" chapter video - somewhat 6th minute.

Reply:
I think I know exactly what you are referring to, and I saw that part as well, but with this way to work, we wouldn't have to initilze the bias in the first place. Ofc we could do that and just put the Bias right from the beginning into the weight matrics. I personally like to compute it separately, because for me it is more clear that there is a bias, which needs no input. Otherwise, we would have to append the input matric always with a magical 1. I don't like that so much in a readability way, but we can do it that way if you want

Reply:
Nono, let's keep it how you did it - I totally agree on readability. Asking those questions to fill in some potential gaps in my understanding.